provider/awssd/aws_sd.go

```diff
@@ -332,6 +556,123 @@ func (p *AWSSDProvider) submitDeletes(namespaces []*sd.NamespaceSummary, changes
 	return nil
 }
 
+// srvChangesHostnameToIP, for each endpoint of type SRV, converts a single host-based target
+// (as passed to ApplyChanges in some cases for SRV records)
+// to multiple IP-based targets (as supported by AWS servicediscovery for SRV records).
+// This is needed because awssd supports IP-based target but not canonical host-based target when service is an SRV record.
+// https://docs.aws.amazon.com/cloud-map/latest/api/API_RegisterInstance.html#cloudmap-RegisterInstance-request-Attributes
+func (p *AWSSDProvider) srvChangesHostnameToIP(namespaces []*sd.NamespaceSummary, changes *plan.Changes) error {
+	applyConversion := func(changes []*endpoint.Endpoint, convertFn srvEpHostnameToIP) error {
+		chByNsID := p.changesByNamespaceID(namespaces, changes)
+		for nsID, changeList := range chByNsID {
+			services, err := p.ListServicesByNamespaceID(aws.String(nsID))
+			if err != nil {
+				return err
+			}
+			for _, change := range changeList {
+				err := convertFn(change, services)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	}
+
+	err := applyConversion(changes.Create, p.srvNewEpHostnameToIP)
+	if err != nil {
+		return err
+	}
+	err = applyConversion(changes.UpdateNew, p.srvNewEpHostnameToIP)
+	if err != nil {
+		return err
+	}
+	err = applyConversion(changes.UpdateOld, p.srvOldEpHostnameToIP)
+	if err != nil {
+		return err
+	}
+	err = applyConversion(changes.Delete, p.srvOldEpHostnameToIP)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
I am little lost here. I believe the task is to group corresponding A and SRV endpoints in a way that (`A.dnsName == SRV.target`). Each corresponding pair of A and SRV should result in a single instance in the Cloud Map. This can be done purely in a static way without calling the API to list all services.

Maybe it would be useful to create the an abstraction over instance to store all the details (IP, port, SRV target etc) to avoid complicated combining and splitting strings all the time.